superset/db_engine_specs/starrocks.py

```diff
@@ -0,0 +1,269 @@
+        "+ INTERVAL QUARTER({col}) QUARTER - INTERVAL 1 QUARTER",
+        "P1Y": "DATE(DATE_SUB({col}, " "INTERVAL DAYOFYEAR({col}) - 1 DAY))",
+        "1969-12-29T00:00:00Z/P1W": "DATE(DATE_SUB({col}, "
+        "INTERVAL DAYOFWEEK(DATE_SUB({col}, "
+        "INTERVAL 1 DAY)) - 1 DAY))",
+    }
+
+    type_code_map: Dict[int, str] = {}  # loaded from get_datatype only if needed
+
+    custom_errors: Dict[Pattern[str], Tuple[str, SupersetErrorType, Dict[str, Any]]] = {
+        CONNECTION_ACCESS_DENIED_REGEX: (
+            __('Either the username "%(username)s" or the password is incorrect.'),
+            SupersetErrorType.CONNECTION_ACCESS_DENIED_ERROR,
+            {"invalid": ["username", "password"]},
+        ),
+        CONNECTION_INVALID_HOSTNAME_REGEX: (
+            __('Unknown MySQL server host "%(hostname)s".'),
+            SupersetErrorType.CONNECTION_INVALID_HOSTNAME_ERROR,
+            {"invalid": ["host"]},
+        ),
+        CONNECTION_HOST_DOWN_REGEX: (
+            __('The host "%(hostname)s" might be down and can\'t be reached.'),
+            SupersetErrorType.CONNECTION_HOST_DOWN_ERROR,
+            {"invalid": ["host", "port"]},
+        ),
+        CONNECTION_UNKNOWN_DATABASE_REGEX: (
+            __('Unable to connect to database "%(database)s".'),
+            SupersetErrorType.CONNECTION_UNKNOWN_DATABASE_ERROR,
+            {"invalid": ["database"]},
+        ),
+        SYNTAX_ERROR_REGEX: (
+            __(
+                'Please check your query for syntax errors near "%(server_error)s". '
+                "Then, try running your query again."
+            ),
+            SupersetErrorType.SYNTAX_ERROR,
+            {},
+        ),
+    }
+
+    @classmethod
+    def convert_dttm(
+        cls, target_type: str, dttm: datetime, db_extra: Optional[Dict[str, Any]] = None
+    ) -> Optional[str]:
+        sqla_type = cls.get_sqla_column_type(target_type)
+
+        if isinstance(sqla_type, types.Date):
+            return f"STR_TO_DATE('{dttm.date().isoformat()}', '%Y-%m-%d')"
+        if isinstance(sqla_type, types.DateTime):
+            datetime_formatted = dttm.isoformat(sep=" ", timespec="microseconds")
+            return f"""STR_TO_DATE('{datetime_formatted}', '%Y-%m-%d %H:%i:%s.%f')"""
+        return None
+
+    @classmethod
+    def adjust_database_uri(
+        cls, uri: URL, selected_schema: Optional[str] = None
+    ) -> URL:
+        # # TODO: jiqian
+        # uri = uri.set(drivername="mysql")
+        database = uri.database
+        if selected_schema and database:
+            selected_schema = parse.quote(selected_schema, safe="")
+            if "." in database:
+                database = database.split(".")[0] + "." + selected_schema
+            else:
+                database += "." + selected_schema
+            uri = uri.set(database=database)
+
+        return uri
+
+    @classmethod
+    def get_datatype(cls, type_code: Any) -> Optional[str]:
+        if not cls.type_code_map:
```

Comment:
Setting the `type_code_map` in a get method seems somewhat atypical. Could this logic reside elsewhere?

Reply:
for lazy load here

Reply:
I understand the concept of lazy loading, but wondered if this could be defined elsewhere. That said it does seem like this pattern is used for other connectors. This code block (and others) does feel like a copy-and-paste from the MySQL connector and thus (per my previous comment) I wonder if it makes more sense to inherit from `MySQLEngineSpec` and just override what is necessary.

@villebro this is somewhat akin to the Presto/Trino refactor where one could debate they're either one in the same or completely different.

Reply:
Thanks for your suggestion.

I refactored these codes, inherit from `MySQLEngineSpec`. Only the necessary code is kept now.

PTAL

@john-bodley